### DISCUSSION TITLE
can a Stripe checkout session be initiated for anonymous user?

### DISCUSSION THREAD

**Discussion by desert-dionysus** [pay-rails/pay#1049] [Q&A]
My goal is to create a Stripe checkout session for an anonymous user to subscribe, then use the customer information collected during checkout to build a User in Rails via the return_url, but everything I've looked at seems to imply that I must use `current_user.payment_processor.checkout` to instantiate the session with Pay gem -- I suppose I can just fall back to the classic Stripe API but curious if there's a proper way to do this with the gem?

Thanks!

**Comment by excid3** [comment]
You can use the Stripe API directly if you want anonymous checkouts.

**Reply by desert-dionysus** [reply]
Thanks for the prompt response, I figured as much and went ahead with that approach last night.  Thanks for the awesome gem!

**Reply by marckohlbrugge** [reply]
I'm trying to decide on whether to force visitors to first create an account or not.

Can you comment on which functionality/data we lose out on using this approach?

I started to write the following pseudo code, but I realized we will not automatically get the charges or payment methods synced to our database.

```ruby
# checkout#new
session = Stripe::Checkout::Session.create(…)
redirect_to session.url, allow_other_host: true
```

```ruby
# checkout#create
session = Stripe::Checkout::Session.retrieve(params[:session_id])
user = User.create!(email: session.customer_email)
user.create_pay_customer default: true, processor: :stripe, processor_id: session.customer

# sync subscriptions (if any were started using the checkout)
user.pay_customer.sync_with_processor

# FIXME: Does not sync charges or payment methods.
```

**Reply by desert-dionysus** [reply]
@marckohlbrugge you raise an interesting point, I'll need to go back and ensure that I'm syncing the payment method with the approach I took... in our case, we have a free trial so there are no charges, but I'm fairly certain the checkout session response contains the payment intent id which can be used to retrieve charge_id
